Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -165,24 +179,24 @@ Writer Writer::Wrap(std::vector<uint8_t>* out_bytes) {
 // make it a bit easier to review. Refactor these to group the related methods
 // together (probably within their own file rather than here).
 
-void Writer::WriteTag(pb_wire_type_t wiretype, uint32_t field_number) {
-  bool status = pb_encode_tag(&stream_, wiretype, field_number);
-  if (!status) {
-    // TODO(rsgowman): figure out error handling
-    abort();
+Status Writer::WriteTag(pb_wire_type_t wiretype, uint32_t field_number) {
+  bool ok = pb_encode_tag(&stream_, wiretype, field_number);
+  if (!ok) {
+    return getInternalError(PB_GET_ERROR(&stream_));
   }
+  return Status::OK();
 }
 
-void Writer::WriteSize(size_t size) {
+Status Writer::WriteSize(size_t size) {
   return WriteVarint(size);
 }
 
-void Writer::WriteVarint(uint64_t value) {
-  bool status = pb_encode_varint(&stream_, value);
-  if (!status) {
-    // TODO(rsgowman): figure out error handling
-    abort();
+Status Writer::WriteVarint(uint64_t value) {
+  bool ok = pb_encode_varint(&stream_, value);
```

Comment:
Optional: this pattern is used often:
```cpp
bool ok = pb_some_fun(...);
if (!ok) {
  return getInternalError(...);
}
return Status::OK();
```

Consider a shorthand like:
```cpp
return MakeStatus(pb_some_fun(...), &stream_);
// ...
Status MakeStatus(bool ok, Stream* stream) {
  return ok ? Status::OK() : Status{InternalError, PB_GET_ERROR(stream)};
}
```

Also, what does `PB_GET_ERROR` return if `ok == true`? It it's a well-defined value like `nullptr`, then maybe `ok` check can be omitted here? And `MakeStatus` could be a member function, leading to:
```cpp
Status Writer::MakeStatus() {
  const char* errmsg = PB_GET_ERROR(&stream_);
  return errmsg ? Status{InternalError, errmsg} : Status::OK();
}
```

Reply:
> Consider a shorthand like:

The error message doesn't always come from the stream. (And even if it did, there's two unrelated stream types, pb_istream_t and pb_ostream_t to worry about, though that's pretty minor). But we could do MakeStatus(bool, char*).

```
Status s = MakeStatus(pb_whatever(...), "This is an error message");
Status s = MakeStatus(pb_whatever(...), PB_GET_ERROR(&stream_));
```

But I don't like:
* that this hides the fact that this creates an *internal* error. (Those are serious errors which shouldn't be accidentally created for something minor.)
* that this jams the error message together with the operation, without obvious context that it's an error message. (Named parameters would completely solve this for me.) 
  `MakeStatus(x, "Is this an error message? Probably, but it might be difficult to tell")`

Maybe a different name?

`MakeStatusOrInternalError(bool, char*)`
(Not really a fan of that either, but can't immediately think of anything better.)

WDYT?

> Also, what does PB_GET_ERROR return if ok == true?

The docs claim a placeholder string. Inspecting the code, that string is: "(none)"

Reply:
Another alternative: Just eliminate the bool variable:

```
if (!pb_do_something(...)) {
  return getInternalErrorOrWhateverWereDoingHereNow();
}
```

Reply:
Perhaps it's a job for two functions?

```cpp
Status RaiseInternalError(const char* message) {
  return Status{InternalError, message);
}
Status Writer::GetCurrentStatus(bool ok) const {
  return ok ? Status::OK() : RaiseInternalError(PB_GET_ERROR(&stream_));
}
```

Eliminating the bool variable would work too.

> The docs claim a placeholder string. Inspecting the code, that string is: "(none)"

Yeah, we can't rely on that. Looking at the [code](https://cs.corp.google.com/piper///depot/google3/third_party/nanopb/pb.h?q=pb_get_error&l=580):
* they claim that code size is reduced if error messages are disabled, though that's probably not worth it;
* the code to return error is `(stream)->errmsg ? (stream)->errmsg : "(none)"`. *Assuming* the behavior of `stream->errmsg` is reliable, then I guess another alternative is:
```cpp
pb_do_something(...);
if (!LastStatus().ok()) return LastStatus();
// ...
Status Writer::LastStatus() const {
  return stream_.errmsg ? Status{InternalError, stream_.errmsg} : Status::OK();
}
```

Reply:
> I feel that this still doesn't work: Sometimes GetCurrentStatus() needs to take in an error message.

My idea was that in such a case, `RaiseInternalError` should be called directly. But if you feel otherwise, I'm fine with that, let's not spend any more time on this point. Perhaps I'll have a better suggestion further down the line.

> It isn't.

But wouldn't it mean that the behavior of `PB_GET_ERROR` is unreliable as well?

> if the calling code initializes the streams properly

Note that value initialization is all that's necessary here:
```cpp
pb_ostream_t s{}; // All members are value-initialized (set to `0` or equivalent)
pb_do_something(&s);
```